// @knit/needle/index.js
/* @flow */

const path = require('path');
const readPkgUp = require('read-pkg-up');

const pkg = (readPkgUp.sync() || {}).pkg || {};

const ROOT_DIR = process.env.PWD || '';
const MOD_STUB = path.join('modules', 'node_modules');
const DIST_STUB = 'dist';
const LIB_STUB = 'lib';
const ES6_STUB = 'es6';
const UMD_STUB = 'umd';
const TESTS_STUB = '__tests__';
const DATA_STUB = 'data';
const WEBPACK_HOST = process.env.HOST || 'localhost';
const WEBPACK_PORT = parseInt(process.env.PORT, 10) || 8080;
const PKG_ENV = ((pkg.knit || {}).needle || {}).env || {};

module.exports = {
  paths: {
    rootDir: ROOT_DIR,
    data: path.join(ROOT_DIR, DATA_STUB),
    modulesStub: MOD_STUB,
    modules: path.join(ROOT_DIR, MOD_STUB),
    dist: path.join(ROOT_DIR, DIST_STUB),
    distStub: DIST_STUB,
    libStub: LIB_STUB,
    es6Stub: ES6_STUB,
    umdStub: UMD_STUB,
    testsStub: TESTS_STUB,
  },
  webpack: {
    host: WEBPACK_HOST,
    port: WEBPACK_PORT,
    public: `http://${WEBPACK_HOST}:${WEBPACK_PORT}/`,
  },
  proxy: process.env.PROXY || 'http://localhost:5000',
  env: Object.assign({},
    {
      'process.env': {
        NODE_ENV: JSON.stringify(process.env.NODE_ENV),
        BABEL_ENV: JSON.stringify(process.env.BABEL_ENV),
        TEST_ENV: JSON.stringify(process.env.TEST_ENV),
      },
      DEBUG: parseInt(process.env.DEBUG, 10) === 1,
      BASE: JSON.stringify(process.env.BASE) || '',
    },
    Object.keys(PKG_ENV).reduce((acc, k) => (
      Object.assign({}, acc, {
        [k]: JSON.stringify(process.env[k] || PKG_ENV[k]),
      })
    ), {})
  ),
  pkg,
};
