// @knit/mittens-develop/index.js
/* @flow weak */

exports.command = 'dev';
exports.describe = 'built-in developer tools';
exports.builder = (ys) => (
  ys
  .command('server', 'start a dev server',
    (yargs) => (
      yargs
        .options({
          p: {
            alias: 'port',
            description: 'set server port',
            type: 'number',
          },
          h: {
            alias: 'host',
            description: 'set server host',
            type: 'string',
          },
          r: {
            alias: 'proxy',
            description: 'set proxy uri',
            type: 'string',
          },
        })
    ), require('./cli-play'))
  .command('test', 'run tests',
    (yargs) => (
      yargs
        .usage('knit dev test -- [jestOptions]')
        .example('knit dev test -- --watch', 'run tests in watch mode')
  ), require('./cli-test'))
  .command('lint', 'run linter', yargs => yargs.help(), require('./cli-lint'))
  .command('flow', 'run flow', yargs => yargs.help(), require('./cli-flow'))
  .command('schema', 'update graphql schema', yargs => yargs.help(), require('./cli-relay'))
);
