// @knit/knit-core/index.js
/* @flow */

import type { TPkgJson, TPaths } from '@knit/needle';

import readPkg from '@knit/read-pkg';

export type TModules = Array<string>;


type TGetDependencyVersion = (d: string, m: {internal: TModules, updated: TModules}, p: { nextVersion: string, paths: TPaths, rootPkg: TPkgJson}) => string;
export const getDependencyVersion: TGetDependencyVersion = (dep, modules, params) => {
  if (params.rootPkg.dependencies && params.rootPkg.dependencies[dep]) {
    return params.rootPkg.dependencies[dep];
  } else if (modules.updated.includes(dep)) {
    return params.nextVersion;
  } else if (modules.internal.includes(dep)) {
    const pkg = readPkg(dep, params.paths);
    if (pkg && pkg.version) {
      return pkg.version;
    }
  }

  throw {
    message: `Missing dependency: ${dep}`,
    stderr: `Could not find ${dep} in the project package.json. Try \`yarn add ${dep}\`.`,
  };
};

type TGetPeerDependencyVersion = (d: string, m: {internal: TModules, updated: TModules}, p: { nextVersion: string, paths: TPaths, pkg: TPkgJson, rootPkg: TPkgJson}) => string;
export const getPeerDependencyVersion: TGetPeerDependencyVersion = (dep, modules, params) => {
  const version = (params.pkg.peerDependencies && params.pkg.peerDependencies[dep])
    ? params.pkg.peerDependencies[dep]
    : params.rootPkg.peerDependencies[dep];

  if (version === '*') {
    const v = (params.rootPkg.devDependencies && params.rootPkg.devDependencies[dep])
      ? params.rootPkg.devDependencies[dep]
      : getDependencyVersion(dep, modules, params);

    return parseInt(v.replace(/^[^\d]/, ''), 10).toString();
  }

  return version;
};

type TGetOptionalDependencyVersion = (d: string, m: {internal: TModules, updated: TModules}, p: { nextVersion: string, paths: TPaths, pkg: TPkgJson, rootPkg: TPkgJson}) => string;
export const getOptionalDependencyVersion: TGetOptionalDependencyVersion = (dep, modules, params) => {
  const version = (params.pkg.optionalDependencies && params.pkg.optionalDependencies[dep])
    ? params.pkg.optionalDependencies[dep]
    : params.rootPkg.optionalDependencies[dep];

  if (version === '*') {
    return getDependencyVersion(dep, modules, params);
  }

  return version;
};

type TUpdateModulePkg = (m: {internal: TModules, updated: TModules, used: TModules }, p: { paths: TPaths, pkg: TPkgJson, rootPkg: TPkgJson, version: ?string }) => TPkgJson;
export const updateModulePkg: TUpdateModulePkg = (modules, params) => {
  const deps = modules.used.concat(Object.keys(params.pkg.dependencies || {}));
  const peers = Object.keys(params.pkg.peerDependencies || {}).concat(
    Object.keys(params.rootPkg.peerDependencies || {})
  );
  const opts = Object.keys(params.pkg.optionalDependencies || {}).concat(
    Object.keys(params.rootPkg.optionalDependencies || {})
  );
  const nextVersion = params.version || params.rootPkg.version;

  return ({
    homepage: params.rootPkg.homepage,
    license: params.rootPkg.license,
    bugs: params.rootPkg.bugs,
    tags: params.rootPkg.tags,
    keywords: params.rootPkg.keywords,
    files: params.rootPkg.files,
    repository: params.rootPkg.repository,
    engines: params.rootPkg.engines,
    author: params.rootPkg.author,
    private: false,
    ...params.pkg,
    ...{
        // fall back to project package if no version given in workflow
      version: nextVersion,
      dependencies: deps.filter(d => !peers.includes(d) && !opts.includes(d)).reduce((acc, d) => ({
        ...acc,
        [d]: getDependencyVersion(d, modules, { ...params, nextVersion }),
      }), {}),
      peerDependencies: peers.reduce((acc, d) => ({
        ...acc,
        [d]: getPeerDependencyVersion(d, modules, { ...params, nextVersion }),
      }), {}),
      optionalDependencies: opts.reduce((acc, d) => ({
        ...acc,
        [d]: getOptionalDependencyVersion(d, modules, { ...params, nextVersion }),
      }), {}),
    },
  });
};
