// @knit/knit/lib/build.js
/* @flow weak */

const path = require('path');
const webpack = require('webpack');
const webpackMerge = require('webpack-merge');
const execa = require('execa');
const upperCamelCase = require('uppercamelcase');

const needle = require('@knit/needle');
const knit = require('@knit/knit-core');

const buildModule = (module, output, options) => {
  const moduleDir = path.join(needle.paths.modules, module);
  return execa('babel', [
    moduleDir,
    '-d', path.join(moduleDir, needle.paths.distStub, output),
    '--ignore', [needle.paths.distStub, needle.paths.testsStub].join(','),
  ], options);
};

exports.buildModuleLib = (module) => (
   buildModule(
    module,
    needle.paths.libStub,
    { env: Object.assign({}, process.env, { NODE_ENV: 'production' }) }
  )
);

exports.buildModuleEs6 = (module) => (
   buildModule(
    module,
    needle.paths.jsnextStub,
    { env: Object.assign({}, process.env, { JS_NEXT: true, NODE_ENV: 'production' }) }
  )
);

exports.buildModuleUmd = (module) => {
  const moduleDir = path.join(needle.paths.modules, module);

  return new Promise((resolve, reject) => webpack(webpackMerge(
    require('@knit/webpack-config-socks-lib'),
    {
      entry: path.join(moduleDir, 'index.js'),
      output: {
        library: upperCamelCase(knit.getModuleName(module)),
        path: path.join(moduleDir, needle.paths.distStub, needle.paths.umdStub),
        filename: 'index.min.js',
      },
    }
  ), (err) => {
    if (err) reject(err);
    resolve();
  }));
};

exports.buildProjectUmd = (module) => (
  new Promise((resolve, reject) => webpack(webpackMerge(
    require('@knit/webpack-config-socks-app'),
    {
      entry: path.join(needle.paths.modules, module, 'index.js'),
      output: {
        path: path.join(needle.paths.dist, 'js'),
      },
    }
  ), (err) => {
    if (err) reject(err);
    resolve();
  }))
);
