// @knit/webpack-config-socks/stage.js
/* @flow */

const webpack = require('webpack');
const findCacheDir = require('find-cache-dir');
const CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin');

const needle = require('@knit/needle');

module.exports = {
  devtool: '#source-map',
  module: {
    preLoaders: [],
    loaders: [
      {
        test: /\.js$/,
        loader: 'babel',
        include: needle.paths.modules,
        query: {
          babelrc: false,
          presets: [require('@knit/babel-preset-socks')],
          cacheDirectory: findCacheDir({
            name: '@knit/webpack-config-socks',
          }),
        },
      },
      {
        test: /\.css$/,
        loader: 'style!css',
      },
      {
        test: /\.json$/,
        loader: 'json',
      },
      {
        test: /\.(ico|jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2)(\?.*)?$/,
        loader: 'file',
        query: {
          name: '[name].[hash:8].[ext]',
        },
      },
      {
        test: /\.(mp4|webm|wav|mp3|m4a|aac|oga)(\?.*)?$/,
        loader: 'url',
        query: {
          limit: 10000,
          name: '[name].[hash:8].[ext]',
        },
      },
    ],
    noParse: [/\.min\.js$/],
  },
  resolve: {
    extentions: ['', 'js', 'json'],
    modules: ['node_modules'],
  },
  plugins: [
    new webpack.DefinePlugin(needle.env),
    new CaseSensitivePathsPlugin(),
  ],
};
