// @knit/jest-config-socks/index.js
/* @flow */

const path = require('path');
const needle = require('@knit/needle');

module.exports = () => {
  const jestConfig = {
    cacheDirectory: '.jest',
    collectCoverageFrom: [
      path.join(needle.paths.rootDir, '**', needle.paths.testsStub, '**', '*.js'),
    ],
    moduleNameMapper: {
      '^[./a-zA-Z0-9$_-]+\\.(jpg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm)$': path.resolve(__dirname, 'fileMock.js'),
      '^[./a-zA-Z0-9$_-]+\\.css$': path.resolve(__dirname, 'styleMock.js'),
    },
    scriptPreprocessor: path.resolve(__dirname, 'transform.js'),
    testEnvironment: 'jsdom',
    testPathDirs: [needle.paths.modules],
    // need to reset all the ignore paths to clear out node_modules
    coveragePathIgnorePatterns: [],
    preprocessorIgnorePatterns: [],
    testPathIgnorePatterns: [],
    haste: {
      // we nee this because haste auto ignores node_modules
      providesModuleNodeModules: ['.*'],
    },
  };

  return Object.assign({}, jestConfig, needle.pkg.jest || {});
};
